File: sdk/go/superdev.go
// SuperDev Go SDK
// go get github.com/superdev/sdk

package superdev

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

const (
	DefaultBaseURL = "http://localhost:8000"
	DefaultTimeout = 60 * time.Second
)

type Config struct {
	BaseURL  string
	APIKey   string
	Timeout  time.Duration
	HTTPClient *http.Client
}

type Client struct {
	config    Config
	httpClient *http.Client
	baseURL   *url.URL
}

func NewClient(config Config) (*Client, error) {
	if config.BaseURL == "" {
		config.BaseURL = DefaultBaseURL
	}
	if config.Timeout == 0 {
		config.Timeout = DefaultTimeout
	}

	baseURL, err := url.Parse(config.BaseURL)
	if err != nil {
		return nil, fmt.Errorf("invalid base URL: %w", err)
	}

	httpClient := config.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: config.Timeout}
	}

	return &Client{
		config:     config,
		httpClient: httpClient,
		baseURL:    baseURL,
	}, nil
}

func (c *Client) request(ctx context.Context, method, path string, body interface{}, result interface{}) error {
	var bodyReader io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("failed to marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(jsonBody)
	}

	fullURL := c.baseURL.ResolveReference(&url.URL{Path: path})
	req, err := http.NewRequestWithContext(ctx, method, fullURL.String(), bodyReader)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	if c.config.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.APIKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		return fmt.Errorf("API error: %d - %s", resp.StatusCode, string(respBody))
	}

	if result != nil {
		if err := json.Unmarshal(respBody, result); err != nil {
			return fmt.Errorf("failed to unmarshal response: %w", err)
		}
	}

	return nil
}

// Health Check
func (c *Client) HealthCheck(ctx context.Context) (*HealthResponse, error) {
	var result HealthResponse
	err := c.request(ctx, "GET", "/api/v1/health", nil, &result)
	return &result, err
}

func (c *Client) GetVersion(ctx context.Context) (*VersionResponse, error) {
	var result VersionResponse
	err := c.request(ctx, "GET", "/api/v1/version", nil, &result)
	return &result, err
}

// Chat
func (c *Client) Chat(ctx context.Context, req ChatRequest) (*ChatResponse, error) {
	var result ChatResponse
	err := c.request(ctx, "POST", "/api/v1/chat/completions", req, &result)
	return &result, err
}

// Code Verification
func (c *Client) VerifyCode(ctx context.Context, req VerificationRequest) (*VerificationResult, error) {
	var result VerificationResult
	err := c.request(ctx, "POST", "/api/v1/verify", req, &result)
	return &result, err
}

// Workflows
func (c *Client) CreateWorkflow(ctx context.Context, req CreateWorkflowRequest) (*WorkflowResponse, error) {
	var result WorkflowResponse
	err := c.request(ctx, "POST", "/api/v1/workflows", req, &result)
	return &result, err
}

func (c *Client) ExecuteWorkflow(ctx context.Context, workflowID string, variables map[string]interface{}) (map[string]interface{}, error) {
	var result map[string]interface{}
	body := map[string]interface{}{"variables": variables}
	err := c.request(ctx, "POST", fmt.Sprintf("/api/v1/workflows/%s/execute", workflowID), body, &result)
	return result, err
}

// Knowledge Base
func (c *Client) CreateKnowledgeBase(ctx context.Context, req KnowledgeBaseCreate) (*KnowledgeBase, error) {
	var result KnowledgeBase
	err := c.request(ctx, "POST", "/api/v1/knowledge-bases", req, &result)
	return &result, err
}

func (c *Client) SearchKnowledge(ctx context.Context, req SearchRequest) (*SearchResponse, error) {
	var result SearchResponse
	err := c.request(ctx, "POST", "/api/v1/knowledge-bases/search", req, &result)
	return &result, err
}

// Projects
func (c *Client) CreateProject(ctx context.Context, req CreateProjectRequest) (*Project, error) {
	var result Project
	err := c.request(ctx, "POST", "/api/v1/projects", req, &result)
	return &result, err
}

func (c *Client) ListProjects(ctx context.Context) ([]Project, error) {
	var result struct {
		Success bool     `json:"success"`
		Data    []Project `json:"data"`
	}
	err := c.request(ctx, "GET", "/api/v1/projects", nil, &result)
	return result.Data, err
}

// Types

type HealthResponse struct {
	Success bool   `json:"success"`
	Data    struct {
		Status    string                 `json:"status"`
		Version   string                 `json:"version"`
		Environment string               `json:"environment"`
		Checks    map[string]interface{} `json:"checks"`
	} `json:"data"`
}

type VersionResponse struct {
	Success bool `json:"success"`
	Data    struct {
		Version string `json:"version"`
		Name    string `json:"name"`
	} `json:"data"`
}

type ChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type ChatRequest struct {
	Messages    []ChatMessage `json:"messages"`
	Model       string        `json:"model,omitempty"`
	Provider    string        `json:"provider,omitempty"`
	Temperature float64       `json:"temperature,omitempty"`
	MaxTokens   int           `json:"max_tokens,omitempty"`
	Stream      bool          `json:"stream,omitempty"`
}

type ChatResponse struct {
	ID          string         `json:"id"`
	Content     string         `json:"content"`
	Model       string         `json:"model"`
	Usage       map[string]interface{} `json:"usage"`
	FinishReason string        `json:"finish_reason"`
}

type VerificationRequest struct {
	TaskDescription string            `json:"task_description"`
	Language        string            `json:"language,omitempty"`
	Context         string            `json:"context,omitempty"`
	Requirements    []string          `json:"requirements,omitempty"`
	ExistingCode    string            `json:"existing_code,omitempty"`
	TestFiles       map[string]string `json:"test_files,omitempty"`
	MaxIterations   int               `json:"max_iterations,omitempty"`
	Provider        string            `json:"provider,omitempty"`
}

type VerificationResult struct {
	TaskID       string                 `json:"task_id"`
	Success      bool                   `json:"success"`
	Stage        string                 `json:"stage"`
	FinalCode    string                 `json:"final_code,omitempty"`
	Error        string                 `json:"error,omitempty"`
	Iterations   int                    `json:"iterations"`
	Generation   map[string]interface{} `json:"generation,omitempty"`
	Execution    map[string]interface{} `json:"execution,omitempty"`
	Testing      map[string]interface{} `json:"testing,omitempty"`
	Review       map[string]interface{} `json:"review,omitempty"`
	Correction   map[string]interface{} `json:"correction,omitempty"`
}

type WorkflowStep struct {
	ID              string                 `json:"id,omitempty"`
	Name            string                 `json:"name"`
	StepType        string                 `json:"step_type"`
	Config          map[string]interface{} `json:"config,omitempty"`
	DependsOn       []string               `json:"depends_on,omitempty"`
	MaxRetries      int                    `json:"max_retries,omitempty"`
	TimeoutSeconds  int                    `json:"timeout_seconds,omitempty"`
	ContinueOnError bool                   `json:"continue_on_error,omitempty"`
}

type CreateWorkflowRequest struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	Steps       []WorkflowStep         `json:"steps"`
	Variables   map[string]interface{} `json:"variables,omitempty"`
	Tags        []string               `json:"tags,omitempty"`
}

type WorkflowResponse struct {
	ID          string                 `json:"id"`
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	Steps       []interface{}          `json:"steps"`
	Tags        []string               `json:"tags"`
}

type KnowledgeBaseCreate struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Type        string `json:"type,omitempty"`
	IsPublic    bool   `json:"is_public,omitempty"`
}

type KnowledgeBase struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Type        string `json:"type"`
	IsPublic    bool   `json:"is_public"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
}

type SearchRequest struct {
	Query                 string   `json:"query"`
	KnowledgeBaseIDs      []string `json:"knowledge_base_ids,omitempty"`
	TopK                  int      `json:"top_k,omitempty"`
	SimilarityThreshold   float64  `json:"similarity_threshold,omitempty"`
}

type SearchResult struct {
	EntryID     string  `json:"entry_id"`
	ChunkID     string  `json:"chunk_id"`
	Title       string  `json:"title"`
	Content     string  `json:"content"`
	Similarity  float64 `json:"similarity"`
	Language    string  `json:"language,omitempty"`
	Tags        []string `json:"tags"`
}

type SearchResponse struct {
	Results []SearchResult `json:"results"`
	Total   int            `json:"total"`
}

type CreateProjectRequest struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Template    string `json:"template,omitempty"`
}

type Project struct {
	ID        string                 `json:"id"`
	Name      string                 `json:"name"`
	Description string               `json:"description"`
	CreatedAt string                 `json:"created_at"`
	UpdatedAt string                 `json:"updated_at"`
	OwnerID   string                 `json:"owner_id"`
	Settings  map[string]interface{} `json:"settings"`
}

File: sdk/rust/superdev.rs
// SuperDev Rust SDK
// cargo add superdev-sdk

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::time::Duration;

#[derive(Debug, Clone)]
pub struct Config {
    pub base_url: String,
    pub api_key: Option<String>,
    pub ws_url: Option<String>,
    pub timeout: Duration,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            base_url: "http://localhost:8000".to_string(),
            api_key: None,
            ws_url: None,
            timeout: Duration::from_secs(60),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Client {
    config: Config,
    client: reqwest::Client,
}

impl Client {
    pub fn new(config: Config) -> Result<Self, Box<dyn std::error::Error>> {
        let client = reqwest::Client::builder()
            .timeout(config.timeout)
            .build()?;

        Ok(Self { config, client })
    }

    async fn request<T: for<'de> Deserialize<'de>>(
        &self,
        method: reqwest::Method,
        path: &str,
        body: Option<&impl Serialize>,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let url = format!("{}{}", self.config.base_url, path);
        let mut request = self.client.request(method, &url);

        if let Some(api_key) = &self.config.api_key {
            request = request.bearer_auth(api_key);
        }

        if let Some(body) = body {
            request = request.json(body);
        }

        let response = request.send().await?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await?;
            return Err(format!("API error: {} - {}", status, text).into());
        }

        let result = response.json().await?;
        Ok(result)
    }

    // Health Check
    pub async fn health_check(&self) -> Result<HealthResponse, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::GET, "/api/v1/health", None::<&()>).await
    }

    // Version
    pub async fn get_version(&self) -> Result<VersionResponse, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::GET, "/api/v1/version", None::<&()>).await
    }

    // Chat
    pub async fn chat(&self, req: ChatRequest) -> Result<ChatResponse, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, "/api/v1/chat/completions", Some(&req)).await
    }

    // Code Verification
    pub async fn verify_code(&self, req: VerificationRequest) -> Result<VerificationResult, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, "/api/v1/verify", Some(&req)).await
    }

    // Workflows
    pub async fn create_workflow(&self, req: CreateWorkflowRequest) -> Result<WorkflowResponse, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, "/api/v1/workflows", Some(&req)).await
    }

    pub async fn execute_workflow(&self, workflow_id: &str, variables: HashMap<String, serde_json::Value>) -> Result<serde_json::Value, Box<dyn std::error::Error>> {
        let body = serde_json::json!({ "variables": variables });
        self.request(reqwest::Method::POST, &format!("/api/v1/workflows/{}/execute", workflow_id), Some(&body)).await
    }

    // Knowledge Base
    pub async fn create_knowledge_base(&self, req: KnowledgeBaseCreate) -> Result<KnowledgeBase, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, "/api/v1/knowledge-bases", Some(&req)).await
    }

    pub async fn search_knowledge(&self, req: SearchRequest) -> Result<SearchResponse, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, "/api/v1/knowledge-bases/search", Some(&req)).await
    }

    // Projects
    pub async fn create_project(&self, req: CreateProjectRequest) -> Result<Project, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, "/api/v1/projects", Some(&req)).await
    }

    pub async fn list_projects(&self) -> Result<Vec<Project>, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::GET, "/api/v1/projects", None::<&()>).await
    }

    // Agents
    pub async fn list_agents(&self) -> Result<Vec<Agent>, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::GET, "/api/v1/agents", None::<&()>).await
    }

    pub async fn execute_agent_task(&self, agent_id: &str, req: ExecuteAgentRequest) -> Result<serde_json::Value, Box<dyn std::error::Error>> {
        self.request(reqwest::Method::POST, &format!("/api/v1/agents/{}/execute", agent_id), Some(&req)).await
    }

    // Plugins
    pub async fn list_plugins(&self, plugin_type: Option<&str>, tag: Option<&str>, search: Option<&str>) -> Result<Vec<Plugin>, Box<dyn std::error::Error>> {
        let mut path = "/api/v1/plugins/registry".to_string();
        let mut params = vec![];
        if let Some(t) = plugin_type {
            params.push(format!("plugin_type={}", t));
        }
        if let Some(t) = tag {
            params.push(format!("tag={}", t));
        }
        if let Some(t) = search {
            params.push(format!("search={}", t));
        }
        if !params.is_empty() {
            path.push('?');
            path.push_str(&params.join("&"));
        }
        self.request(reqwest::Method::GET, &path, None::<&()>).await
    }
}

// Types

#[derive(Debug, Deserialize)]
pub struct HealthResponse {
    pub status: String,
    pub version: String,
    pub environment: String,
    pub timestamp: String,
}

#[derive(Debug, Deserialize)]
pub struct VersionResponse {
    pub version: String,
    pub name: String,
}

#[derive(Debug, Serialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Serialize)]
pub struct ChatRequest {
    pub messages: Vec<ChatMessage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct ChatResponse {
    pub id: String,
    pub content: String,
    pub model: String,
    pub usage: serde_json::Value,
    pub finish_reason: String,
}

#[derive(Debug, Serialize)]
pub struct VerificationRequest {
    pub task_description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub requirements: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub existing_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub test_files: Option<HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_iterations: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct VerificationResult {
    pub task_id: String,
    pub success: bool,
    pub stage: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub final_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    pub iterations: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generation: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub testing: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub review: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correction: Option<serde_json::Value>,
}

#[derive(Debug, Serialize)]
pub struct WorkflowStep {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub name: String,
    pub step_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub depends_on: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_retries: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout_seconds: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub continue_on_error: Option<bool>,
}

#[derive(Debug, Serialize)]
pub struct CreateWorkflowRequest {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub steps: Vec<WorkflowStep>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct WorkflowResponse {
    pub id: String,
    pub name: String,
    pub description: String,
    pub steps: Vec<serde_json::Value>,
    pub tags: Vec<String>,
}

#[derive(Debug, Serialize)]
pub struct KnowledgeBaseCreate {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_public: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct KnowledgeBase {
    pub id: String,
    pub name: String,
    pub description: String,
    pub r#type: String,
    pub is_public: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize)]
pub struct SearchRequest {
    pub query: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub knowledge_base_ids: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_k: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub similarity_threshold: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct SearchResult {
    pub entry_id: String,
    pub chunk_id: String,
    pub title: String,
    pub content: String,
    pub similarity: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    pub tags: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct SearchResponse {
    pub results: Vec<SearchResult>,
    pub total: i32,
}

#[derive(Debug, Serialize)]
pub struct CreateProjectRequest {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub template: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Project {
    pub id: String,
    pub name: String,
    pub description: String,
    pub created_at: String,
    pub updated_at: String,
    pub owner_id: String,
    pub settings: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Deserialize)]
pub struct Agent {
    pub id: String,
    pub name: String,
    pub r#type: String,
    pub status: String,
    pub config: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Serialize)]
pub struct ExecuteAgentRequest {
    pub task: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Deserialize)]
pub struct Plugin {
    pub name: String,
    pub slug: String,
    pub version: String,
    pub description: String,
    pub author: String,
    pub plugin_type: String,
    pub tags: Vec<String>,
    pub downloads: i32,
    pub rating: f64,
    pub is_official: bool,
}